hudi-client/hudi-spark-client/src/main/java/org/apache/hudi/index/bloom/SparkHoodieMetadataBloomIndexHelper.java

```diff
@@ -0,0 +1,167 @@
+ * under the License.
+ */
+
+package org.apache.hudi.index.bloom;
+
+import org.apache.hudi.avro.model.HoodieMetadataColumnStats;
+import org.apache.hudi.common.data.HoodieData;
+import org.apache.hudi.common.data.HoodiePairData;
+import org.apache.hudi.common.engine.HoodieEngineContext;
+import org.apache.hudi.common.fs.FSUtils;
+import org.apache.hudi.common.model.HoodieKey;
+import org.apache.hudi.common.model.HoodieRecordLocation;
+import org.apache.hudi.common.util.HoodieTimer;
+import org.apache.hudi.common.util.collection.ImmutablePair;
+import org.apache.hudi.common.util.collection.Pair;
+import org.apache.hudi.config.HoodieWriteConfig;
+import org.apache.hudi.data.HoodieJavaPairRDD;
+import org.apache.hudi.data.HoodieJavaRDD;
+import org.apache.hudi.exception.HoodieMetadataException;
+import org.apache.hudi.exception.MetadataNotFoundException;
+import org.apache.hudi.index.HoodieIndexUtils;
+import org.apache.hudi.io.HoodieKeyLookupResult;
+import org.apache.hudi.table.HoodieTable;
+import org.apache.log4j.LogManager;
+import org.apache.log4j.Logger;
+import org.apache.spark.api.java.JavaRDD;
+import org.apache.spark.api.java.function.Function2;
+import scala.Tuple2;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+import java.util.stream.Collectors;
+
+import static java.util.stream.Collectors.toList;
+
+/**
+ * Helper for {@link HoodieBloomIndex} containing Spark-specific logic.
+ */
+public class SparkHoodieMetadataBloomIndexHelper extends MetadataBaseHoodieBloomIndexHelper {
+
+  private static final Logger LOG = LogManager.getLogger(SparkHoodieMetadataBloomIndexHelper.class);
+  private static final SparkHoodieMetadataBloomIndexHelper SINGLETON_INSTANCE =
+      new SparkHoodieMetadataBloomIndexHelper();
+
+  private SparkHoodieMetadataBloomIndexHelper() {
+  }
+
+  public static SparkHoodieMetadataBloomIndexHelper getInstance() {
+    return SINGLETON_INSTANCE;
+  }
+
+  protected HoodieData<Pair<String, IndexFileFilter>> buildPartitionIndexFileFilter(
+      HoodieWriteConfig config, final HoodieEngineContext context, final HoodieTable hoodieTable,
+      HoodiePairData<String, String> partitionRecordKeyPairs) {
+    final String keyField = hoodieTable.getMetaClient().getTableConfig().getRecordKeyFieldProp();
+    final int inputParallelism = HoodieJavaPairRDD.getJavaPairRDD(partitionRecordKeyPairs).partitions().size();
+    final int joinParallelism = Math.max(inputParallelism, config.getBloomIndexParallelism());
+    LOG.info("Building partition index file filter, input parallelism: " + inputParallelism
+        + ", bloom index parallelism: " + config.getBloomIndexParallelism());
+    JavaRDD<String> partitionsRDD = HoodieJavaPairRDD.getJavaPairRDD(partitionRecordKeyPairs).keys().distinct().repartition(joinParallelism);
+
+    JavaRDD<Pair<String, IndexFileFilter>> partitionIndexFileFilterRDD = partitionsRDD.mapPartitionsWithIndex(
```

Comment:
curious to understand why we have gone w/ mapPartitionsWithIndex. here we are iterating through a list of partitions. thought map would suffice. likely the parallelism(spark partitions) will be equal to num partitions.